compiler/resolution/resolveFunction.cpp

```diff
@@ -984,6 +984,109 @@ static void computeReturnTypeParamVectors(BaseAST*      ast,
                                           Vec<Type*>&   retTypes,
                                           Vec<Symbol*>& retSymbols);
 
+void resolveIfExprType(CondStmt* stmt) {
+  Vec<Type*> retTypes;
+  Vec<Symbol*> retSymbols;
+
+  CallExpr* last = toCallExpr(stmt->thenStmt->body.tail);
+  INT_ASSERT(last->isPrimitive(PRIM_MOVE));
+  Symbol* ret = toSymExpr(last->get(1))->symbol();
+  FnSymbol* fn = toFnSymbol(stmt->parentSymbol); // dummy to use for canDispatch
+
+  computeReturnTypeParamVectors(stmt, ret, retTypes, retSymbols);
+
+  if (retTypes.n == 1) {
+    ret->type = retTypes.head();
+  } else {
+    INT_ASSERT(retTypes.n == 2 && retSymbols.n == 2);
+
+    // NB: Assumes '0' and '1' correspond to 'then' and 'else', respectively.
+    Type* thenType = retTypes.v[0];
+    Type* elseType = retTypes.v[1];
+    Type* retType  = NULL;
+
+    Symbol* thenSym = retSymbols.v[0];
+    Symbol* elseSym = retSymbols.v[1];
+
+    bool thenTypeVar = thenSym->hasFlag(FLAG_TYPE_VARIABLE);
+    bool elseTypeVar = elseSym->hasFlag(FLAG_TYPE_VARIABLE);
+
+    // If only one branch returns a reference, that branch needs to copy the
+    // result so that we can always free the result.
+    if (isReferenceType(thenType) != isReferenceType(elseType)) {
+      BlockStmt* refBranch = isReferenceType(thenType) ? stmt->thenStmt : stmt->elseStmt;
+      CallExpr* call = toCallExpr(refBranch->body.tail);
+      SymExpr* rhs = toSymExpr(call->get(2));
+      if (isUserDefinedRecord(rhs->getValType())) {
+        CallExpr* copy = new CallExpr("chpl__autoCopy", rhs->remove());
+        call->insertAtTail(copy);
+        resolveCallAndCallee(copy);
+        if (isReferenceType(thenType)) {
+          thenType = copy->resolvedFunction()->retType;
+        } else {
+          elseType = copy->resolvedFunction()->retType;
+        }
+      }
+
+      thenType = thenType->getValType();
+      elseType = elseType->getValType();
+    }
+
+    if (thenType == elseType) {
+      retType = thenType;
+    } else {
+      bool promote = false;
+
+      if (canDispatch(elseType, elseSym, thenType, fn, &promote) &&
+          promote == false) {
+        retType = thenType;
+      } else if (canDispatch(thenType, thenSym, elseType, fn, &promote) &&
+                 promote == false) {
+        retType = elseType;
+      }
+    }
+
+    // For tuples, generally do not allow a tuple to contain a reference
+    // when it is returned
+    if (retType != NULL && retType->symbol->hasFlag(FLAG_TUPLE) == true) {
```

Comment:
I don't think we need this tuple case here - it should be handled by normal variable initialization?

Try removing it and check what the type of x is in:
``` chapel
var A:[1..10] int;
var B:[1..10] int;
var x = if something then (A,B) else (B,A);
```

x should have a tuple type storing the arrays by value, but the if-expr can return a tuple type storing the arrays by reference.

Reply:
I originally omitted this case, but ran into a testing failure and so copied it from the resolve-return-type function.

Reply:
That's odd, I wonder if something else is wrong / the test needs updating. It's at least worth trying the example I showed above and looking at the --log output to see if we need to make an issue about it.

Reply:
The case in which a program would behave differently is like this:
``` chapel
f( if something then (A,B) else (B,A) );
```

Reply:
I have opened #9684 for further discussion of this issue.